// tsutil/either.js
var Optional = require("./optional");
/**
 * Either wraps a left or right value, but not both.
 * To represent a computation that is an Error/Result, use Try.
 */
var Either = (function () {
    /**
     * You must provide left xor right.
     *
     * That is one of left and right must be defined and not null,
     * but not both of them.
     * @param left
     * @param right
     */
    function Either(left, right) {
        this._left = undefined;
        this._right = undefined;
        var leftDefined = Either._isDefined(left);
        var rightDefined = Either._isDefined(right);
        if (!leftDefined && !rightDefined) {
            throw new Error("Either#constructor(): Left or right must be defined");
        }
        else if (leftDefined && rightDefined) {
            throw new Error("Either#constructor(): Left and right cannot both be defined");
        }
        if (leftDefined) {
            this._left = left;
        }
        else {
            this._right = right;
        }
    }
    /**
     * Returns an Either with the provided left value.
     * @param left
     * @returns {*}
     */
    Either.left = function (left) {
        return new Either(left, undefined);
    };
    /**
     * Returns an Either with the provided right value.
     * @param right
     * @returns {*}
     */
    Either.right = function (right) {
        return new Either(undefined, right);
    };
    Either._isDefined = function (value) {
        return typeof value !== "undefined";
    };
    /**
     * Calls and returns the result of onLeft on the left if it's defined,
     * otherwise calls and returns the result of onRight on the right.
     * @param onLeft
     * @param onRight
     * @returns {*}
     */
    Either.prototype.fold = function (onLeft, onRight) {
        if (this.isLeft()) {
            return onLeft(this._left);
        }
        else {
            return onRight(this._right);
        }
    };
    /**
     * Whether left is defined.
     * @returns {*}
     */
    Either.prototype.isLeft = function () {
        return Either._isDefined(this._left);
    };
    /**
     * Whether right is defined.
     * @returns {*}
     */
    Either.prototype.isRight = function () {
        return Either._isDefined(this._right);
    };
    /**
     * Returns an Optional over left, or Optional.NONE if it is not defined.
     * @returns {*}
     */
    Either.prototype.left = function () {
        return new Optional(this._left);
    };
    /**
     * Returns left or null if it is not defined.
     * Use of this method is discouraged. Prefer using use #left(), #fold(),
     * etc. instead.
     * @returns {*}
     */
    Either.prototype.leftOrNull = function () {
        if (this.isLeft()) {
            return this._left;
        }
        else {
            return null;
        }
    };
    /**
     * Returns an Optional over right, or Optional.NONE if it is not defined.
     * @returns {*}
     */
    Either.prototype.right = function () {
        return new Optional(this._right);
    };
    /**
     * Returns right or null if it is not defined.
     * Use of this method is discouraged. Prefer using use #right(), #fold(),
     * etc. instead.
     * @returns {*}
     */
    Either.prototype.rightOrNull = function () {
        if (this.isRight()) {
            return this._right;
        }
        else {
            return null;
        }
    };
    /**
     * Returns a new Either whose right is this' left,
     * and whose left is this' right.
     * @returns {*}
     */
    Either.prototype.swap = function () {
        return new Either(this._right, this._left);
    };
    return Either;
})();
module.exports = Either;
